stacker/dag/__init__.py

```diff
@@ -168,22 +173,27 @@ def transitive_reduction(self):
 
         See https://en.wikipedia.org/wiki/Transitive_reduction
         """
-
-        graph = self.graph
-        for x in graph.keys():
-            for y in graph.keys():
-                for z in graph.keys():
-                    # Edge from x -> y
-                    xy = y in graph[x]
-                    # Edge from y -> x
-                    yz = z in graph[y]
-                    # Edge from x -> z
-                    xz = z in graph[x]
-
-                    # If edges xy and yz exist, remove edge xz.
-                    if xz and xy and yz:
-                        logger.debug("Removing edge %s -> %s" % (x, z))
-                        graph[x].remove(z)
+        combinations = []
```

Comment:
ping @ejholmes  :)

Reply:
@phobologic Hi Michael, thanks for you interest in this new DAG implementation. As I've stated in f6733af3, the previous implementation is flawed in that it cannot handle some of the corner cases.

Take a real world example here, software A depends on B and D, while B depends on C and C depends on D. What this `transitive_reduction` does is to simply the dependency tree to `A -> B -> C -> D`, while the previous implementation cannot return the correct result.

As for the implemetation, the idea is with all current edges(`A -> B`, `A -> D`, `B -> C`, `C -> D`), we will join the edges to create a new edge if the end of one edge is the beginning of another, so for `A -> B` and `B -> C`, we'll create `A -> C`, and, since we are running in a loop, we'll use `A -> C` and `C -> D` to create `A -> D` and all that. After this loop, we have marked all the connectivity between nodes. Then we'll mark all those newly constructed edges and remove the unnecessary nodes.

I've discovered this flaw during a Python 3.5 test run, where the key order of dictionaries are different from Python 2.7 and Python 3.6.

As for performance, correct me if I'm wrong but I think both implementation are O(N^3), and I don't believe this could be an issue since we are using DAG to handle Cloudformation templates, which will never be huge.